lib/Sema/TypeCheckType.cpp

```diff
@@ -810,21 +810,26 @@ static Type diagnoseUnknownType(TypeChecker &tc, DeclContext *dc,
           (nominal = nominalDC->getAsNominalTypeOrNominalTypeExtensionContext())) {
         // Attempt to refer to 'Self' within a non-protocol nominal
         // type. Fix this by replacing 'Self' with the nominal type name.
-
-        // Retrieve the nominal type and resolve it within this context.
         assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
-        auto type = resolver->mapTypeIntoContext(
-          dc->getInnermostTypeContext()->getSelfInterfaceType());
-        if (type->hasError())
-          return type;
 
         // Produce a Fix-It replacing 'Self' with the nominal type name.
         auto name = getDeclNameFromContext(dc, nominal);
         tc.diagnose(comp->getIdLoc(), diag::self_in_nominal, name)
           .fixItReplace(comp->getIdLoc(), name);
-        comp->overwriteIdentifier(nominal->getName());
-        comp->setValue(nominal, nominalDC->getParent());
-        return type;
+
+        // If this is a requirement, replacing 'Self' with a valid type will
+        // result in additional unnecessary diagnostics (does not refer to a
+        // generic parameter or associated type). Simply return an error type.
+        if (options.contains(TypeResolutionFlags::GenericRequirement)) {
+          return ErrorType::get(tc.Context);
+        } else {
```

Comment:
As a code style thing, swiftc typically prefers early returns and avoiding `else`s/indentation in cases like this:

```c++
if (options.contains(TypeResolutionFlags::GenericRequirement))
  return ErrorType::get(tc.Context);

auto type = resolver->mapTypeIntoContext(
  dc->getInnermostTypeContext()->getSelfInterfaceType());

...
```

Reply:
Ah, that was an overlook